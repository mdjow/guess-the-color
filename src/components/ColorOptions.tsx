import styled from "styled-components";

interface ColorOptionsProps {
  options: string[];
  disabled: boolean;
  onSelect: (color: string) => void;
}

export const ColorOptions = ({
  onSelect,
  options,
  disabled,
}: ColorOptionsProps) => (
  <ColorOptionsWrapper>
    {options.map((color, i) => (
      <Option
        key={i}
        onClick={() => onSelect(color)}
        color={color}
        disabled={disabled}
      >
        {color}
      </Option>
    ))}
  </ColorOptionsWrapper>
);

const ColorOptionsWrapper = styled.div`
  display: flex;
  gap: 0.3rem;
`;

const Option = styled.button`
  width: 100%;
  padding: 1rem;
  font-size: 1rem;
  font-weight: 600;
  border-radius: var(--border-radius);
  cursor: pointer;
  background-color: #ffffff;
  border: 1px solid #e0e0e0;

  &:hover:enabled {
    background-color: #eeeeee;
  }
`;
