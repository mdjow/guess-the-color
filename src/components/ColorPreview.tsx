import styled from "styled-components";

interface ColorPreviewProps {
  color: string;
  started: boolean;
  onStart: () => void;
}

export const ColorPreview = ({
  color,
  started,
  onStart,
}: ColorPreviewProps) => (
  <ColorPreviewWrapper color={color}>
    {!started && <StartButton onClick={onStart}>Start</StartButton>}
  </ColorPreviewWrapper>
);

const ColorPreviewWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 20rem;
  background-color: ${(props) => props.color};
`;

const StartButton = styled.button`
  font-size: 1.5rem;
  padding: 0.7rem 3rem;
  border-radius: var(--border-radius);
  cursor: pointer;
  background-color: #fff;
  border: 1px solid #e0e0e0;
`;
