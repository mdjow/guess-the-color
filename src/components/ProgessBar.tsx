import styled from "styled-components";

interface ProgressBarProps {
  progress: number;
}

export const ProgressBar = styled.div<ProgressBarProps>`
  position: relative;
  height: 0.6rem;
  background-color: #f5f5f5;

  &::after {
    content: "";
    display: flex;
    width: ${(props) => props.progress}%;
    height: 100%;
    background-color: #4caf50;
    transition: width 1s linear;
    background-color: 0.2s ease;
  }
`;
