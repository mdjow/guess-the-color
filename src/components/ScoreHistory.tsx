import styled from "styled-components";

type ScoreHistoryProps = {
  timeRemaining: number;
  score: number;
  highScore: number;
  onReset: () => void;
};

export const ScoreHistory = (props: ScoreHistoryProps) => (
  <ScoreHistoryWrapper>
    <BoxRow>
      <span>Remaning Time (s):</span>
      <span>{props.timeRemaining}</span>
    </BoxRow>
    <RestartButton onClick={props.onReset}>Restart</RestartButton>
    <BoxColumn>
      <BoxRow>
        <span>High score:</span>
        <span>{props.highScore}</span>
      </BoxRow>
      <BoxRow>
        <span>Score:</span>
        <span>{props.score}</span>
      </BoxRow>
    </BoxColumn>
  </ScoreHistoryWrapper>
);

const ScoreHistoryWrapper = styled.div`
  display: flex;
  width: 100%;
  background-color: #f5f5f5;
  border: 1px solid #e0e0e0;
  border-radius: 0.2rem;

  > *:not(:last-child) {
    border-right: 1px solid #e0e0e0;
  }
`;

const BoxColumn = styled.div`
  display: flex;
  flex-direction: column;
  font-weight: 300;
  width: 100%;
`;

const BoxRow = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  box-sizing: border-box;

  width: 100%;
  padding: 0.2rem 0.5rem;
  font-weight: 400;

  :nth-child(2) {
    font-weight: 700;
    font-size: 1.4rem;
  }

  &:nth-child(2) {
    border-top: 1px solid #e0e0e0;
  }
`;

const RestartButton = styled.button`
  width: 100%;
  padding: 0.3rem;
  font-size: 1.2rem;
  border: none;
  cursor: pointer;
  background-color: #e0e0e0;

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
`;
