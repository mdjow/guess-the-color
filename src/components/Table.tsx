import styled from "styled-components";

interface CellProps {
  colspan?: number;
}

export const Table = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Row = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 1rem 0;
  border-bottom: 1px solid var(--border-color);
`;

export const Cell = styled.div<CellProps>`
  flex: ${(props) => props.colspan ?? 1};
  text-align: center;
  padding: 0 0.5rem;
`;
