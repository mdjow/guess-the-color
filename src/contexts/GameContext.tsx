/* eslint-disable react-refresh/only-export-components */
import { ReactNode, createContext, useReducer } from "react";

import { gameReducer } from "./gameReducer";

const TIME_GAME_REMAINING = 30000;

export interface GameAnswer {
  pick: string;
  color: string;
  score: number;
}

export interface GameState {
  started: boolean;
  timeGameRemaining: number;
  currentColor: string;
  colorOptions: string[];
  score: number;
  highScore: number;
  answers: GameAnswer[];
}

export interface GameContextType {
  state: GameState;
  startGame: () => void;
  startNewRound: () => void;
  pickColor: (color: string) => void;
  decrementTime: () => void;
  endGame: () => void;
  resetGame: () => void;
  resetData: () => void;
}

export const initialState: GameState = {
  started: false,
  timeGameRemaining: TIME_GAME_REMAINING,
  currentColor: "#eeeeee",
  colorOptions: ["#######", "#######", "######"],
  score: 0,
  highScore: parseInt(localStorage.getItem("highScore") || "0"),
  answers: JSON.parse(localStorage.getItem("answers") || "[]"),
};

export const GameContext = createContext<GameContextType>({
  state: initialState,
  startGame: () => {
    throw new Error("not Implemented");
  },
  startNewRound: () => {
    throw new Error("not Implemented");
  },
  pickColor: () => {
    throw new Error("not Implemented");
  },
  decrementTime: () => {
    throw new Error("not Implemented");
  },
  endGame: () => {
    throw new Error("not Implemented");
  },
  resetGame: () => {
    throw new Error("not Implemented");
  },
  resetData: () => {
    throw new Error("not Implemented");
  },
});

export const GameProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer(gameReducer, initialState);

  const startGame = () => {
    dispatch({ type: "START_GAME" });
    dispatch({ type: "START_NEW_ROUND" });
  };

  const startNewRound = () => {
    dispatch({ type: "START_NEW_ROUND" });
  };

  const pickColor = (color: string) => {
    dispatch({ type: "PICK_COLOR", color });
    dispatch({ type: "START_NEW_ROUND" });
  };

  const decrementTime = () => {
    dispatch({ type: "DECREMENT_TIME" });
  };

  const resetGame = () => {
    dispatch({ type: "RESET_GAME" });
  };

  const endGame = () => {
    dispatch({ type: "END_GAME" });
  };

  const resetData = () => {
    dispatch({ type: "RESET_DATA" });
  };

  return (
    <GameContext.Provider
      value={{
        state,
        startGame,
        startNewRound,
        pickColor,
        decrementTime,
        endGame,
        resetGame,
        resetData,
      }}
    >
      {children}
    </GameContext.Provider>
  );
};
