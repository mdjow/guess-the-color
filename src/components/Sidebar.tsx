import { useContext } from "react";

import styled from "styled-components";
import { CheckCircleOutlined, HighlightOff } from "@mui/icons-material";

import { TableCell, TableRow, Table, TableBody, TableHeader } from "./Table";
import { calculateTextColor } from "../utils/calculateTextColor";
import { GameAnswer, GameContext } from "../context/gameContext";
import { ScoreEnum } from "../enum/score";

export const Sidebar = () => {
  const { state } = useContext(GameContext);

  const buildColorCells = (answer: GameAnswer) => {
    if (!answer.pick) {
      return (
        <TableCell colSpan={2}>
          <TimeOutBox>Time Out</TimeOutBox>
        </TableCell>
      );
    }
    if (answer.pick === answer.color) {
      return (
        <TableCell colSpan={2}>
          <ColorBox color={answer.pick} />
        </TableCell>
      );
    }

    return (
      <>
        <TableCell>
          <ColorBox color={answer.pick} />
        </TableCell>
        <TableCell>
          <ColorBox color={answer.color} />
        </TableCell>
      </>
    );
  };

  return (
    <SidebarWrapper>
      <Title> Current/Latest Game</Title>
      <Table>
        <TableHeader>
          <TableCell>Guessed color</TableCell>
          <TableCell>Correct color</TableCell>
          <TableCell>Score</TableCell>
        </TableHeader>
        <TableBody>
          {state?.answers.map((answer) => (
            <TableRow key={`${answer.pick}-${answer.color}`}>
              {buildColorCells(answer)}
              <TableCell>
                <Score>
                  {answer.score === ScoreEnum.SUCCESS ? (
                    <CheckCircleOutlined color="success" />
                  ) : (
                    <HighlightOff color="error" />
                  )}
                  {answer.score}
                </Score>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </SidebarWrapper>
  );
};

const SidebarWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 20rem;
  border-right: 1px solid var(--border-color);
`;

const Title = styled.h1`
  font-size: 1.5rem;
  font-weight: 600;
  text-align: center;
`;

interface ColorBoxProps {
  color?: string;
}

const ColorBox = styled.div<ColorBoxProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 0.2rem;
  border-radius: var(--border-radius);
  background-color: ${(props) => props.color};

  &::after {
    content: "${(props) => props.color}";
    color: ${(props) => props.color && calculateTextColor(props.color)};
  }
`;

const TimeOutBox = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 0.2rem;
  border-radius: var(--border-radius);
  border: 1px solid #e0e0e0;
`;

const Score = styled.div`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  font-weight: 600;
`;
