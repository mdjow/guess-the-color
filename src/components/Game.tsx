import { useContext, useEffect, useRef } from "react";

import styled from "styled-components";

import { ColorOptions } from "./ColorOptions";
import { ColorPreview } from "./ColorPreview";
import { ProgressBar } from "./ProgessBar";
import { ScoreHistory } from "./ScoreHistory";
import { GameContext } from "../contexts/GameContext";

const TIME_ANSWER_REMAINING = 10000;

export const Game = () => {
  const {
    state,
    startGame,
    resetData,
    resetGame,
    pickColor,
    decrementTime,
    endGame,
  } = useContext(GameContext);

  const {
    started,
    score,
    highScore,
    currentColor,
    colorOptions,
    timeGameRemaining,
  } = state;

  const timerGameRef = useRef<NodeJS.Timeout>();
  const timerAnswerRef = useRef<NodeJS.Timeout>();

  useEffect(() => {
    if (timeGameRemaining === 0) {
      clearInterval(timerGameRef.current);
      clearInterval(timerAnswerRef.current);
      endGame();
    }
  }, [endGame, timeGameRemaining]);

  const setTimerAnswerInterval = () => {
    if (timerAnswerRef.current) {
      clearInterval(timerAnswerRef.current);
    }

    timerAnswerRef.current = setInterval(
      () => pickColor(""),
      TIME_ANSWER_REMAINING
    );
  };

  const handleStartGame = () => {
    startGame();

    timerGameRef.current = setInterval(() => {
      decrementTime();
    }, 1000);

    setTimerAnswerInterval();
  };

  const handleResetGame = () => {
    clearInterval(timerGameRef.current);
    clearInterval(timerAnswerRef.current);

    resetGame();
  };

  const handleResetData = () => {
    clearInterval(timerGameRef.current);
    clearInterval(timerAnswerRef.current);

    resetData();
  };

  const handlePickColor = (color: string) => {
    setTimerAnswerInterval();

    pickColor(color);
  };

  return (
    <GameWrapper>
      <Title>Guess the color</Title>
      <ScoreHistory
        timeRemaining={timeGameRemaining}
        score={score}
        highScore={highScore}
        disabled={!started}
        onReset={handleResetGame}
      />
      <ProgressBar progress={timeGameRemaining / 300} />
      <ColorPreview
        color={currentColor}
        onStart={handleStartGame}
        started={started}
      />
      <ColorOptions
        options={colorOptions}
        disabled={!started}
        onSelect={handlePickColor}
      />
      <ResetDataButton data-testid="reset-all-data" onClick={handleResetData}>
        Reset all data
      </ResetDataButton>
    </GameWrapper>
  );
};

const GameWrapper = styled.main`
  display: flex;
  flex-direction: column;
  margin: auto;
  gap: 1.3rem;
  width: 40rem;
  padding: 3rem;
`;

const Title = styled.h1`
  font-size: 2.5rem;
  font-weight: 800;
  text-align: center;
`;

const ResetDataButton = styled.a`
  position: absolute;
  bottom: 1rem;
  right: 1rem;
  cursor: pointer;
  text-decoration: underline;
`;
