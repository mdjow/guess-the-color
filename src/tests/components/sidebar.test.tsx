import "jest-styled-components";

import { render } from "@testing-library/react";

import { GameContext } from "../../contexts/GameContext";
import { Sidebar } from "../../components/Sidebar";
import { createMockGameContext } from "../mocks/game.mock";

describe("Sidebar component", () => {
  test("should render component matches snapshot", () => {
    const mockContext = createMockGameContext();

    const mockState = {
      ...mockContext.state,
      answers: [
        { pick: "#FC2703", color: "#FC2703", score: 5 },
        { pick: "#337FF2", color: "#9F22E0", score: -1 },
        { pick: "", color: "#FD6D4A", score: -2 },
      ],
    };

    const { container } = render(
      <GameContext.Provider value={{ ...mockContext, state: mockState }}>
        <Sidebar />
      </GameContext.Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
