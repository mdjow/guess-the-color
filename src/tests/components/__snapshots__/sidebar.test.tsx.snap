// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component should render component matches snapshot 1`] = `
.c2 {
  display: flex;
  flex-direction: column;
  overflow-y: auto;
}

.c3 {
  display: flex;
  justify-content: space-between;
  padding: 1rem 0;
  border-bottom: 1px solid var(--border-color);
  font-weight: 600;
}

.c5 {
  display: flex;
  flex-direction: column;
  overflow-y: auto;
}

.c6 {
  display: flex;
  justify-content: space-between;
  padding: 1rem 0;
  border-bottom: 1px solid var(--border-color);
}

.c4 {
  flex: 1;
  text-align: center;
  padding: 0 0.5rem;
}

.c7 {
  flex: 2;
  text-align: center;
  padding: 0 0.5rem;
}

.c0 {
  display: flex;
  flex-direction: column;
  width: 20rem;
  border-right: 1px solid var(--border-color);
}

.c1 {
  font-size: 1.5rem;
  font-weight: 600;
  text-align: center;
}

.c8 {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 0.2rem;
  border-radius: var(--border-radius);
  background-color: #FC2703;
}

.c8::after {
  content: "#FC2703";
  color: #ffffff;
}

.c10 {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 0.2rem;
  border-radius: var(--border-radius);
  background-color: #337FF2;
}

.c10::after {
  content: "#337FF2";
  color: #ffffff;
}

.c11 {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 0.2rem;
  border-radius: var(--border-radius);
  background-color: #9F22E0;
}

.c11::after {
  content: "#9F22E0";
  color: #ffffff;
}

.c12 {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 0.2rem;
  border-radius: var(--border-radius);
  border: 1px solid #e0e0e0;
}

.c9 {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  font-weight: 600;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
     Current/Latest Game
  </h1>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        Guessed color
      </div>
      <div
        class="c4"
      >
        Correct color
      </div>
      <div
        class="c4"
      >
        Score
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        <div
          class="c7"
          colspan="2"
        >
          <div
            class="c8"
            color="#FC2703"
          />
        </div>
        <div
          class="c4"
        >
          <div
            class="c9"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium css-4b67si-MuiSvgIcon-root"
              data-testid="CheckCircleOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm4.59-12.42L10 14.17l-2.59-2.58L6 13l4 4 8-8z"
              />
            </svg>
            5
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c4"
        >
          <div
            class="c10"
            color="#337FF2"
          />
        </div>
        <div
          class="c4"
        >
          <div
            class="c11"
            color="#9F22E0"
          />
        </div>
        <div
          class="c4"
        >
          <div
            class="c9"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-1wxstni-MuiSvgIcon-root"
              data-testid="HighlightOffIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.59 8 12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
            -1
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
          colspan="2"
        >
          <div
            class="c12"
          >
            Time Out
          </div>
        </div>
        <div
          class="c4"
        >
          <div
            class="c9"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-1wxstni-MuiSvgIcon-root"
              data-testid="HighlightOffIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.59 8 12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
            -2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
